cmd/podman/images/scan.go

```diff
@@ -0,0 +1,158 @@
+package images
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/containers/common/pkg/completion"
+	"os"
+	"text/template"
+
+	"github.com/containers/podman/v2/cmd/podman/common"
+	"github.com/containers/podman/v2/cmd/podman/registry"
+	"github.com/containers/podman/v2/libpod/define"
+	"github.com/containers/podman/v2/pkg/domain/entities"
+	"github.com/containers/podman/v2/pkg/specgen"
+	"github.com/spf13/cobra"
+)
+
+var (
+	imageScanCommand = &cobra.Command{
+		Use:     "scan [run-options] IMAGE [scanner-options]",
+		Short:   "Scan a container image for known vulnerabilities",
+		Long:    `Scan a container image for known vulnerabilities. The image name or digest can be used.`,
+		Args:    cobra.MinimumNArgs(1),
+		RunE:    scan,
+		Example: `podman image scan centos:latest`,
+	}
+
+	scanOptions entities.ImageScanOptions
+)
+
+func scanFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+
+	flagName := "scanner"
+	flags.StringVarP(
+		&scanOptions.ScannerImage,
+		flagName, "s", "docker.io/anchore/syft:latest",
+		"The vulnerability scanner container image to use",
+	)
+	_ = cmd.RegisterFlagCompletionFunc(flagName, completion.AutocompleteNone)
+
+	flagName = "mount-point"
+	flags.StringVarP(
+		&scanOptions.MountPoint,
+		flagName, "m", "/podman-scan-image-mount",
+		"Mount-point of the target image to be scanned",
+	)
+	_ = cmd.RegisterFlagCompletionFunc(flagName, completion.AutocompleteNone)
+
+	flagName = "env"
+	flags.StringSliceVarP(
+		&scanOptions.Env,
+		flagName, "e", nil,
+		"Set environment variables in scanner tool container",
+	)
+	_ = cmd.RegisterFlagCompletionFunc(flagName, completion.AutocompleteNone)
+
+}
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Mode:    []entities.EngineMode{entities.ABIMode},
+		Command: imageScanCommand,
+		Parent:  imageCmd,
+	})
+	scanFlags(imageScanCommand)
+}
+
+func scan(cmd *cobra.Command, args []string) error {
+	imageToScan := args[0]
+
+	// render all scanner args with the mount-point (if necessary)
+	type ScanTemplate struct {
+		MountPoint string
+	}
+	values := ScanTemplate{MountPoint: scanOptions.MountPoint}
+	tpl := template.New("podman-scan")
+
+	scannerArgs := append([]string{scanOptions.ScannerImage}, args[1:]...)
+	for i, arg := range scannerArgs {
+		parsedTpl, err := tpl.Parse(arg)
+		if err != nil {
+			continue
+		}
+
+		buf := bytes.NewBufferString("")
+		if err = parsedTpl.Execute(buf, values); err != nil {
+			continue
+		}
+		scannerArgs[i] = buf.String()
+	}
+
+	mountOptions := fmt.Sprintf("type=image,src=%s,target=%s,readwrite=false", imageToScan, scanOptions.MountPoint)
+
+	ctrConfig := registry.PodmanConfig()
+
+	ctrOpts := common.ContainerCLIOpts{
+		Mount:             []string{mountOptions},
+		ImageVolume:       common.DefaultImageVolume,
```

Comment:
@mheon, do we need to set all these options here or does `FillOutSpecGen` take care of it?  I am a bit worried of code duplication and potential divergence in the future.

Reply:
Using the CLI parsing code for this is a bit iffy because it was written to expect CLI-provided defaults for flags.

Reply:
I would say we probably need it as written. Directly using Specgen would remove the necessity.